### PR TITLE
Add portforwarding command

### PR DESCRIPTION
**Pull request by Jefferson-Faseler**
Adds `port-forward` command. Can be used with all features of the `kubectl` command of the same name.
`tuber port-forward -a user-service 8989`
To open a port on localhost:8989 -> pod::8989

Ports can be a variadic list that also take advantage of the `:` syntax in the `kubectl port-forward` [command](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#port-forward). 

`tuber port-forward -a user-service -p user-service-8822272-fsdalhj 9001` to specify a particular pod.

Able to specify the workload, local IP addresses, and pod wait time in addition to the small examples above.

Also, as an added bonus I gave `exec` the ability to specify a pod as well. Specifying a pod has many added benefits, like faster command times and edge-cases for a-typically configured pods.

### CODE REVIEW

File: cmd/helpers.go [hunk 1]
```diff
@@ -13,6 +14,12 @@ import (
 	"github.com/spf13/viper"
 )
 
+var address string
+var appName string
+var pod string
+var podRunningTimeout string
+var workload string
```
[the remark this answers is not in the document]
Comment (jaysonesmith) [reply cmd/helpers.go:21]: Oh wait, are these being set by Cobra?

**Review by jaysonesmith — COMMENTED** [review]
Not sure my reviews are necessarily going to be super useful right now, but I've asked some questions to try to get a better understanding and stuff.

File: cmd/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jefferson-Faseler) [reply cmd/helpers.go:21]: Yes, they are set by Cobra in commands that share the need for these vars; `exec`, and `port-forward`.

File: cmd/portForward.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	ports := args
+
+	portForwardArgs := []string{"--address", address, "--pod-running-timeout", podRunningTimeout}
+	err = k8s.PortForward(podName, appName, ports, portForwardArgs...)
+	return err
+}
+
+func init() {
+	portForwardCmd.Flags().StringVar(&address, "address", "localhost", "specify an address on your local machine to forward from. Can be a comma separated list. Only IP addresses and 'localhost' are valid.")
```
[the remark this answers is not in the document]
Comment (Jefferson-Faseler) [reply cmd/portForward.go:46]: Yeah, let k8s do the work for parsing it. No need to reinvent the wheel here IMO.

File: cmd/exec.go [hunk 3]
```diff
@@ -15,50 +13,29 @@ var execCmd = &cobra.Command{
...
-
-	labels := strings.TrimSuffix(string(l), ",")
-
-	jsonPath := fmt.Sprintf(`-o=jsonpath="%s"`, `{.items[0].metadata.name}`)
-	name, err := k8s.GetCollection("pods", appName, "-l", labels, jsonPath)
-	if err != nil {
-		return err
+		containerName = fetchWorkload()
 	}
+	podName, err := fetchPodname()
```
[the remark this answers is not in the document]
Comment (Jefferson-Faseler) [reply cmd/exec.go:26]: Oh gosh. 🤦 Nice catch

File: cmd/exec.go [hunk 4]
```diff
@@ -15,50 +13,29 @@ var execCmd = &cobra.Command{
...
-		return err
-	}
-
-	labels := strings.TrimSuffix(string(l), ",")
-
-	jsonPath := fmt.Sprintf(`-o=jsonpath="%s"`, `{.items[0].metadata.name}`)
-	name, err := k8s.GetCollection("pods", appName, "-l", labels, jsonPath)
-	if err != nil {
-		return err
+		containerName = fetchWorkload()
```
[the remark this answers is not in the document]
Comment (Jefferson-Faseler) [reply cmd/exec.go:24]: It is also used in https://github.com/Freshly/tuber/blob/add-portforwarding-command/cmd/helpers.go#L196 for fetching pod names.

**Review by jkminneti — CHANGES_REQUESTED** [review]
using the simplest example, 
```
tuber exec -a user-service -w user-service-sidekiq
```
it returns 
```
error: you must specify at least one command for the container
```
upon the kubectl exec.

debug logs indicate the pod name was successfully identified.

**Comment by Jefferson-Faseler** [comment]
> using the simplest example,
> 
> ```
> tuber exec -a user-service -w user-service-sidekiq
> ```
> 
> it returns
> 
> ```
> error: you must specify at least one command for the container
> ```
> 
> upon the kubectl exec.
> 
> debug logs indicate the pod name was successfully identified.

I'm confused. Are you suggesting I make changes to the `exec` command? I think the output would describe why you are seeing the error.

File: cmd/portForward.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	Long: `Forward requests from a local address and port to a running pod.
+You are able to specify multiple addresses and ports, but all combinations must be valid and running.
+
+This command will always run against a single pod until either that pod terminates or this command is closed.
+
+Specifying a workload:
+The workload name will default to the app name if not supplied. If the workload name is not the same as the app name, that argument will be required to run the command successfully.
+
+For example. When the desired workload is a deployment named 'user-service-sidekiq' within a tuber app named 'user-service':
+tuber exec -a user-service -w user-service-sidekiq
```
Comment (jkminneti) [inline cmd/portForward.go:24]: not exec

Comment (Jefferson-Faseler) [reply cmd/portForward.go:24]: Oh hey, good catch

Comment (jkminneti) [reply cmd/portForward.go:24]: i had copy pasted from that line in the docs (to explain my previous comment lol)

File: cmd/portForward.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	"tuber/pkg/k8s"
+
+	"github.com/spf13/cobra"
+)
+
+// portForwardCmd represents the portForward command
+var portForwardCmd = &cobra.Command{
+	SilenceUsage: true,
+	Use:          "port-forward -a appName [port(s)]",
+	Args:         cobra.MinimumNArgs(1),
```
Comment (jkminneti) [inline cmd/portForward.go:13]: exactargs?

Comment (Jefferson-Faseler) [reply cmd/portForward.go:13]: Nope, as many as your ❤️ desires

Comment (Jefferson-Faseler) [reply cmd/portForward.go:13]: Well, technically 65535 args max

Comment (jkminneti) [reply cmd/portForward.go:13]: why though? not like it's variadic

Comment (Jefferson-Faseler) [reply cmd/portForward.go:13]: It is https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#port-forward

Comment (jkminneti) [reply cmd/portForward.go:13]: `one or more local ports to a pod` why would anyone... but whatever you're absolutely right

File: cmd/portForward.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"tuber/pkg/k8s"
+
+	"github.com/spf13/cobra"
+)
+
+// portForwardCmd represents the portForward command
+var portForwardCmd = &cobra.Command{
+	SilenceUsage: true,
+	Use:          "port-forward -a appName [port(s)]",
```
Comment (jkminneti) [inline cmd/portForward.go:12]: ok call it pedantic, but i don't like the `-a`. could easily do `tuber port-forward payment-service 8080`

in fact, what if we assumed 8080 by default

File: cmd/portForward.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+func portForward(cmd *cobra.Command, args []string) error {
+	podName, err := fetchPodname()
+	if err != nil {
+		return err
+	}
+	ports := args
+
+	portForwardArgs := []string{"--address", address, "--pod-running-timeout", podRunningTimeout}
+	err = k8s.PortForward(podName, appName, ports, portForwardArgs...)
+	return err
```
Comment (jkminneti) [inline cmd/portForward.go:42]: this should at least print a clickable link. bonus brownie taco pizza points if it were to open your default web browser to it

Comment (Jefferson-Faseler) [reply cmd/portForward.go:42]: I don't think I'm going to do this change. A brownie taco pizza sounds very unpleasant.

File: cmd/portForward.go (hunk 7 above)
Comment (Jefferson-Faseler) [reply cmd/portForward.go:12]: I'm matching what exec's syntax is so ditching `-a` would change that command as well. 

Supplying port(s) is not difficult for a user to do. I would think it is better to have tuber complain someone did not pass a port so they know it's a thing.

File: cmd/portForward.go (hunk 8 above)
Comment (jkminneti) [reply cmd/portForward.go:42]: not even the clickable link though?

File: cmd/portForward.go (hunk 7 above)
Comment (jkminneti) [reply cmd/portForward.go:12]: for us it makes sense, is intuitive, and is easy. but for someone who doesnt understand what port-forwarding is, some sensible defaults could get them into their sidekiq-web. 

i'm imagining "which port do i need for sidekiq" kinds of questions

File: cmd/portForward.go (hunk 8 above)
Comment (Jefferson-Faseler) [reply cmd/portForward.go:42]: I can look into it 😒

File: cmd/portForward.go (hunk 7 above)
Comment (Jefferson-Faseler) [reply cmd/portForward.go:12]: For sidekiq, sure. What about postgres, redis, memcached, applications (tls and http)? Then the portforwarding doesn't work by default and we have the same question

Comment (jkminneti) [reply cmd/portForward.go:12]: that's the thing, though - this isn't supposed to be a `kubectl` wrapper, it's supposed to be a "service mesh tool for app devs"

**Review by jkminneti — APPROVED** [review]

**Pull request merged**